rocketpy/_encoders.py

```diff
@@ -33,11 +39,153 @@ def default(self, o):
             return float(o)
         elif isinstance(o, np.ndarray):
             return o.tolist()
+        elif isinstance(o, datetime):
+            return [o.year, o.month, o.day, o.hour]
+        elif hasattr(o, "__iter__") and not isinstance(o, str):
+            return list(o)
         elif hasattr(o, "to_dict"):
-            return o.to_dict()
-        # elif isinstance(o, Function):
-        #     return o.__dict__()
-        elif isinstance(o, (Function, types.FunctionType)):
-            return repr(o)
+            encoding = o.to_dict(self.include_outputs)
+            encoding = remove_circular_references(encoding)
+
+            encoding["signature"] = get_class_signature(o)
+
+            return encoding
+
+        elif hasattr(o, "__dict__"):
+            encoding = remove_circular_references(o.__dict__)
+
+            if "rocketpy" in o.__class__.__module__:
+                encoding["signature"] = get_class_signature(o)
+
+            return encoding
+        else:
+            return super().default(o)
+
+
+class RocketPyDecoder(json.JSONDecoder):
+    """Custom JSON decoder for RocketPy objects. It defines how to decode
+    different types of objects from a JSON supported format."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(object_hook=self.object_hook, *args, **kwargs)
+
+    def object_hook(self, obj):
+        if "signature" in obj:
+            signature = obj.pop("signature")
+
+            try:
+                class_ = get_class_from_signature(signature)
+
+                if hasattr(class_, "from_dict"):
+                    return class_.from_dict(obj)
+                else:
+                    # Filter keyword arguments
+                    kwargs = {
+                        key: value
+                        for key, value in obj.items()
+                        if key in class_.__init__.__code__.co_varnames
+                    }
+
+                    return class_(**kwargs)
+            except (ImportError, AttributeError):
+                return obj
         else:
-            return json.JSONEncoder.default(self, o)
+            return obj
+
+
+def get_class_signature(obj):
+    """Returns the signature of a class in the form of a string.
+    The signature is an importable string that can be used to import
+    the class by its module.
+
+    Parameters
+    ----------
+    obj : object
+        Object to get the signature from.
+
+    Returns
+    -------
+    str
+        Signature of the class.
+    """
+    class_ = obj.__class__
+    name = getattr(class_, '__qualname__', class_.__name__)
+
+    return {"module": class_.__module__, "name": name}
+
+
+def get_class_from_signature(signature):
+    """Returns the class by importing its signature.
+
+    Parameters
+    ----------
+    signature : str
+        Signature of the class.
+
+    Returns
+    -------
+    type
+        Class defined by the signature.
+    """
```

Comment:
I still cannot know what is a "class signature", is it the signature of the init class?

what do you have in the "signature" dictionary (which bte the docs says it is a str) exactly?

Reply:
You are right, I changed what is a `signature` a few times during development. I will improve the docs.

Reply:
Done in [c8e1682](https://github.com/RocketPy-Team/RocketPy/pull/679/commits/c8e168290dd97e7a4b4699e7c2c285ceea4a16f3).